[run-benchmarks] Use `ktrace artrace` for profiling browser benchmarks

xw opened the pull request:
#### 4b03459ba7f6a33a8e4b610ca229901d64e07eb7
<pre>
[run-benchmarks] Use `ktrace artrace` for profiling browser benchmarks
<a href="https://bugs.webkit.org/show_bug.cgi?id=259669">https://bugs.webkit.org/show_bug.cgi?id=259669</a>
rdar://113167762

Reviewed by Dewei Zhu.

Up until now, the `--profile` argument in run-benchmarks hasn&apos;t done anything. This patch implements profiling using the `ktrace artrace` command.

The --profile argument has been modified to require a TRACE_TYPE argument, this can be either `full` or `profile`. It&apos;ll start profiling when the test starts, and save the resulting files to the directory specified in --diagnose-directory (or /tmp/ by default).

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
(BenchmarkRunner.__init__):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/browser_driver.py:
(BrowserDriver.profile):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_browser_driver.py:
(OSXBrowserDriver):
(OSXBrowserDriver.profile):
* Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
(config_argument_parser):
(parse_args):
(run_benchmark_plan):
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner.__init__):
(WebServerBenchmarkRunner):

Canonical link: <a href="https://commits.webkit.org/266492@main">https://commits.webkit.org/266492@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4be872f3576650075e34282bf5c408871fe86bee

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13959 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14274 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14609 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15697 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13250 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14043 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16783 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14357 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15918 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14129 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14723 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11825 "Passed tests") | [💥 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16404 "An unexpected error occured. Recent messages:Failed to print configuration") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/14054 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12009 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12586 "Passed tests") | [💥 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/16404 "An unexpected error occured. Recent messages:Failed to print configuration") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13085 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12750 "Passed tests") | [💥 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/16404 "An unexpected error occured. Recent messages:Failed to print configuration") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13297 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-WK2-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12571 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3383 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16903 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13139 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Leelour commented on Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
```diff
@@ -57,15 +57,15 @@ def _construct_subtest_url(self, subtests):
 
     def _run_one_test(self, web_root, test_file, iteration):
-        enable_profiling = True if self._profile_output_dir else False
+        enable_profiling = True if self._profile_output_dir and self._trace_type else False
```
Maybe we can just do
`enable_profiling = self._profile_output_dir and self._trace_type `

Leelour commented on Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_browser_driver.py:
```diff
@@ -41,6 +42,42 @@ def restore_env_after_all_testing(self):
+            if not trace_process:
+                return
+            try:
+                subprocess.call(['sudo', 'pkill', '-2', 'ktrace'])
```
Maybe use `trace_process.pid` instead so that it can be more accurate.

xw replied:
I've tried the following:
`trace_process.send_signal(signal.SIGINT)`
`os.kill(trace_process.pid, signal.SIGINT)`
`subprocess.call(['sudo', 'kill', '-2', str(trace_process.pid)])`
None of these seem to work - only `pkill` does. My suspicion is that the PID reported by `trace_process.pid` and the actual pid of the ktrace process are different.

dewei-zhu replied:
Per discussion, I think this has something to do with invoking ktrace with sudo. I think pkill might be the best way to terminate the process.

Leelour commented on Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_browser_driver.py:
```diff
@@ -41,6 +42,42 @@ def restore_env_after_all_testing(self):
+                _log.error('Failed to quit ktrace.')
+            finally:
+                _log.info('Killing ktrace task.')
+                subprocess.call(['sudo', 'pkill', '-9', 'ktrace'])
```
Ditto

Leelour commented on Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_browser_driver.py:
```diff
@@ -41,6 +42,42 @@ def restore_env_after_all_testing(self):
+            _log.error('`{}` is not a valid type of trace. Defaulting to `profile`.'.format(trace_type))
+            trace_type = 'profile'
+        try:
+            _log.info('Gathering traces')
```
Worth add more logging to tell user to type host machine admin password if get a prompt.

dewei-zhu reviewed (approved):
r=me

xw committed:
[run-benchmarks] Use `ktrace artrace` for profiling browser benchmarks
https://bugs.webkit.org/show_bug.cgi?id=259669
rdar://113167762

Reviewed by Dewei Zhu.

Up until now, the `--profile` argument in run-benchmarks hasn't done anything. This patch implements profiling using the `ktrace artrace` command.

The --profile argument has been modified to require a TRACE_TYPE argument, this can be either `full` or `profile`. It'll start profiling when the test starts, and save the resulting files to the directory specified in --diagnose-directory (or /tmp/ by default).

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
(BenchmarkRunner.__init__):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/browser_driver.py:
(BrowserDriver.profile):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_browser_driver.py:
(OSXBrowserDriver):
(OSXBrowserDriver.profile):
* Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
(config_argument_parser):
(parse_args):
(run_benchmark_plan):
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner.__init__):
(WebServerBenchmarkRunner):

Canonical link: https://commits.webkit.org/266492@main
